plugins/codemirror/dialogs/codemirrordialog.js

```diff
@@ -44,7 +44,9 @@ CKEDITOR.dialog.add('codemirrordialog', function (editor) {
 
 			var editor = dialog.getParentEditor();
 
-			codeMirrorEditor.setValue(editor.getData(true));
+			var editable = editor.editable();
+
+			codeMirrorEditor.setValue(editable.getData());
```

Comment:
What's the difference between getting the data from the `editable` vs getting it from the `editor`?

(Just curious; I don't actually understand how any of this works.)

Reply:
That's a good question which I've also been asking myself ... 

My quick answer would be: there should be no difference; **but** my longer answer would be: 
Given my doubts; it's probably a good opportunity to make sure I'm doing things correctly.

> (Just curious; I don't actually understand how any of this works.)

This  (`codemirror`) plugin, [adds a new "mode"](https://github.com/liferay/liferay-ckeditor/blob/2bdf5767b53df9f807fbc90e5b118091f3a6def7/plugins/codemirror/plugin.js#L10) to ckeditor,  almost exactly like the [`sourcearea`](https://github.com/ckeditor/ckeditor4/blob/eb2d11644a796cd13cd047cd55541430359f9317/plugins/sourcearea/plugin.js) plugin, the only difference is that we're using CodeMirror to make the "source code" editor look a bit nicer  that an simple `textarea`.

I first introduced the `editable` in [`cd36001`](https://github.com/liferay/liferay-ckeditor/commit/cd360019946b67e733d2046e6813df44040ef3af), because of [this issue](https://issues.liferay.com/browse/LPS-127948). Since our `codemirror` plugin is based on the [`sourcearea`](https://github.com/ckeditor/ckeditor4/blob/eb2d11644a796cd13cd047cd55541430359f9317/plugins/sourcearea/plugin.js) plugin (it's actually a mix of `sourcearea` and [`sourcedialog`](https://github.com/ckeditor/ckeditor4/tree/eb2d11644a796cd13cd047cd55541430359f9317/plugins/sourcedialog)), I thought it would make sense to "copy" as much of its behavior as possible **except** that I'm doing it wrong ...

For example: 

[This is](https://github.com/liferay/liferay-ckeditor/blob/2bdf5767b53df9f807fbc90e5b118091f3a6def7/plugins/codemirror/plugin.js#L160) incorrect: it should be `this.status` and not `this.value` (which might be causing this or other bugs)

[This](https://github.com/liferay/liferay-ckeditor/blob/2bdf5767b53df9f807fbc90e5b118091f3a6def7/plugins/codemirror/dialogs/codemirrordialog.js#L247-L254) shouldn't be needed because, when the dialog is opened we create a `textarea` (that we wrap with `CodeMirror`) and [set it's value to `editable.getData`](https://github.com/liferay/liferay-ckeditor/blob/2bdf5767b53df9f807fbc90e5b118091f3a6def7/plugins/codemirror/dialogs/codemirrordialog.js#L47-L49) (which should have the same value as `editor.getData()`) so there's no need to check again in the `onShow` method, at least that what I think (and also what is said in the [documentation](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_savedata.html#retrieving-data-from-ckeditor-4))

All this to say, that I'm going to have another look at this, because even if having the `editable` makes sense in source mode (to fix the LPS mentioned above), calling [`editor.setData` should automatically call the editable's `setData` method](https://github.com/ckeditor/ckeditor4/blob/eb2d11644a796cd13cd047cd55541430359f9317/core/editable.js#L849-L851).

So what I think I should do is:
- Create the `editable` when the source more is registered
- When CodeMirror dispatches it's `change` event, call `editor.setData` with whatever value `codeMirrorEditor` has
- When opening the "preview" dialog make sure `editor.getData()` is up to date

I doubt this answered your question, but it clearly showed me I should re-think this.